netcfgbu/cli/report.py

```diff
@@ -1,90 +1,199 @@
+    TIME_FORMAT = "%Y-%b-%d %I:%M:%S %p"
 
-        self.task_results = defaultdict(list)
+    def __init__(self) -> None:
+        """Initializes a new instance of the Report class."""
+        self.start_ts: Union[None, datetime] = None
+        self.start_tm: float = 0.0
+        self.stop_ts: Union[None, datetime] = None
+        self.stop_tm: float = 0.0
+        self.task_results: dict[bool, list[dict[str, Any]]] = defaultdict(list)
 
-    def start_timing(self):
+    def start_timing(self) -> None:
+        """Starts the timing for the report."""
         self.start_ts = datetime.now()
         self.start_tm = monotonic()
 
-    def stop_timing(self):
+    def stop_timing(self) -> None:
+        """Stops the timing for the report."""
         self.stop_ts = datetime.now()
         self.stop_tm = monotonic()
 
     @property
-    def start_time(self):
+    def start_time(self) -> str:
+        """Returns the formatted start time of the report.
+
+        Raises:
+            ValueError: If the start time has not been set.
+
+        Returns:
+            str: The start time as a formatted string.
+        """
+        if self.start_ts is None:
+            raise ValueError("start_ts is not set")
         return self.start_ts.strftime(self.TIME_FORMAT)
 
     @property
-    def stop_time(self):
+    def stop_time(self) -> str:
+        """Returns the formatted stop time of the report.
+
+        Raises:
+            ValueError: If the stop time has not been set.
+
+        Returns:
+            str: The stop time as a formatted string.
+        """
+        if self.stop_ts is None:
+            raise ValueError("stop_ts is not set")
         return self.stop_ts.strftime(self.TIME_FORMAT)
 
     @property
-    def duration(self):
+    def duration(self) -> float:
+        """Returns the duration between the start and stop times.
+
+        Returns:
+            float: The duration in seconds.
+        """
         return self.stop_tm - self.start_tm
 
-    def print_report(self):
+    def save_report(
+        self,
+        filename: str,
+        headers: list[str],
+        data: list[list[Any]],
+        summary_headers: Union[None, list[str]] = None,
+        summary_data: Union[None, dict[str, dict[str, int]]] = None,
+    ) -> None:
+        """Saves the report data to a CSV file and prints a summary if provided.
+
+        Args:
+            filename: The name of the CSV file to save the report to.
+            headers: The headers for the CSV file.
+            data: The data to save in the CSV file.
+            summary_headers: The headers for the summary table (optional).
+            summary_data: The data for the summary table (optional).
+        """
+        data.sort(key=lambda x: x[0])  # Sorting by the first column (usually 'host')
+
+        with open(filename, "w+", encoding="utf-8") as ofile:
+            wr_csv = csv.writer(ofile)
+            wr_csv.writerow(headers)
+            wr_csv.writerows(data)
```

Comment:
**Suggestion:** To improve the efficiency of the `save_report` method, consider using the CSV module's DictWriter instead of writer, which can simplify the code and potentially increase performance when dealing with dictionaries. [performance, importance: 5]
```suggestion
wr_csv = csv.DictWriter(ofile, fieldnames=headers)
wr_csv.writeheader()
wr_csv.writerows(data)
```